File: codes/contest/leetcode/binary-tree-maximum-path-sum.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <algorithm>
#include <climits>
#include <cstdio>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
};

class Solution {
   public:
    // 返回包含必须root节点的最大值
    // kmax 表示当前已知的最大值
    int trace(TreeNode* root, int* kmax) {
        int R = INT_MIN;
        if (root == NULL) {
            *kmax = R;
            return 0;
        }

        int lkmax = 0;
        int lv = trace(root->left, &lkmax);
        int rkmax = 0;
        int rv = trace(root->right, &rkmax);

        // value包括
        // root, root + lv, root + rv
        // kmax包括
        // lkmax, rkmax, root + lv, root + rv, root + lv + rv, root
        int value = max(0, max(lv, rv)) + root->val;
        *kmax = max(value, max(lv + rv + root->val, max(lkmax, rkmax)));
        return value;
    }
    int maxPathSum(TreeNode* root) {
        int kmax = 0;
        trace(root, &kmax);
        return kmax;
    }
};

File: codes/contest/leetcode/binary-tree-maximum-path-sum.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        INT_MIN = -(1 << 30)

        def walk(root):
            if not root:
                return INT_MIN, 0
            left_sum, left_tree = walk(root.left)
            right_sum, right_tree = walk(root.right)
            root_tree = max(left_tree, right_tree, 0) + root.val
            root_sum = max(left_sum, right_sum, max(left_tree, 0) + max(right_tree, 0) + root.val)
            return root_sum, root_tree

        res, _ = walk(root)
        return res
